src/apps/vpn/addons/state/addonsessionstate.cpp

```diff
@@ -0,0 +1,43 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "addonsessionstate.h"
+
+#include <QHash>
+#include <QJsonValue>
+#include <QObject>
+#include <QString>
+
+#include "addonstatebase.h"
+#include "leakdetector.h"
+
+// static
+AddonSessionState* AddonSessionState::fromManifest(
+    const QJsonObject& manifest) {
+  return new AddonSessionState(AddonStateBase::parseManifest(manifest));
+}
+
+AddonSessionState::AddonSessionState(QHash<QString, QJsonValue> spec)
+    : AddonStateBase(spec) {
+  MZ_COUNT_CTOR(AddonSessionState);
+}
+
+AddonSessionState::~AddonSessionState() { MZ_COUNT_DTOR(AddonSessionState); }
+
+QJsonValue AddonSessionState::getInternal(const QString& key) const {
+  return m_state[key];
```

Comment:
I wonder if m_state should be in the base class. You will need it also for the other implementations. The other impls will have a different propagation algorithm but m_state will probably remain.

Reply:
I had it in the bsae class initially. But I am not too sure it will be required for the other states. Like for local state, will we load from settings to memory on initialize? Or will we always just quuery the settings on `set`/`get`/`clear`? Ifwe do that then `m_state` is not necesssary. We can always move it back there if we realize all of them use it.